clang/lib/DPCT/CallExprRewriterCommon.h

```diff
@@ -1453,6 +1458,116 @@ class CheckCanUseTemplateMalloc {
   }
 };
 
+class CheckCanUseCLibraryMalloc {
+  unsigned AddrArgIdx;
+  unsigned SizeArgIdx;
+
+public:
+  CheckCanUseCLibraryMalloc(unsigned AddrIdx, unsigned SizeIdx)
+      : AddrArgIdx(AddrIdx), SizeArgIdx(SizeIdx) {}
+  bool operator()(const CallExpr *C) {
+    if (!DpctGlobalInfo::isOptimizeMigration()) {
+      return false;
+    }
+    if (C->getNumArgs() <= AddrArgIdx)
+      return false;
+    auto AllocatedExpr = C->getArg(AddrArgIdx);
+    const Expr *AE = nullptr;
+    if (auto CSCE = dyn_cast<CStyleCastExpr>(
+            AllocatedExpr->IgnoreImplicitAsWritten())) {
+      AE = CSCE->getSubExpr()->IgnoreImplicitAsWritten();
+    } else {
+      AE = AllocatedExpr;
+    }
+    auto UO = dyn_cast<UnaryOperator>(AE);
+    if (!UO) {
+      return false;
+    }
+    auto DRE = dyn_cast<DeclRefExpr>(UO->getSubExpr());
+    if (!DRE) {
+      return false;
+    }
+    auto &SM = DpctGlobalInfo::getSourceManager();
+    auto &CT = DpctGlobalInfo::getContext();
+    std::string PtrName = DRE->getNameInfo().getName().getAsString();
+    auto DREDecl = DRE->getDecl();
+    if (!DREDecl || CT.getParents(*DREDecl)[0].get<TranslationUnitDecl>() ||
+        DREDecl->isCXXClassMember()) {
+      return false;
+    }
+    auto PtrScope = findImmediateBlock(DREDecl);
+    if (!PtrScope) {
+      return false;
+    }
+    auto PtrMatcher =
+        findAll(declRefExpr(to(varDecl(hasName(PtrName)))).bind("PtrVar"));
+    auto MatchResult = ast_matchers::match(PtrMatcher, *PtrScope, CT);
+    // For following 3 cases, it's safe to use malloc from c runtime library
+    for (auto &SubResult : MatchResult) {
+      bool HostAccess = false;
+      const DeclRefExpr *PtrDRE = SubResult.getNodeAs<DeclRefExpr>("PtrVar");
+      if (!PtrDRE) {
+        return false;
+      }
+      // Case 1: Array subscript expr and accessed as rvalue
+      if (auto Expr =
+                   DpctGlobalInfo::findAncestor<ArraySubscriptExpr>(PtrDRE)) {
+        auto RValueExpr = DpctGlobalInfo::findAncestor<ImplicitCastExpr>(Expr);
+        if (RValueExpr &&
+            (RValueExpr->getCastKind() == CastKind::CK_LValueToRValue)) {
+          HostAccess = true;
+        }
+      }
+      // Case 2: Dereference expr and accessed as rvalue
+      if (auto Expr =
+                   DpctGlobalInfo::findAncestor<UnaryOperator>(PtrDRE)) {
```

Comment:
Use findParent maybe better, since findAncestor has some corner cases to make the check pass while it shouldn't.